### PR TITLE
JAMES-2442 Integration tests for RemoteDelivery

### PR DESCRIPTION
**Pull request by trantienduchn**
- That an unreachable MX IP will generate a failover to the second MX when direct resolution
- If all (direct resolution) MX IPs are not reachable, a bounce is sent
- what happens when direct resolution to a domain with no mx records? a bounce is sent

### CODE REVIEW

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            .sendMessage(FROM, RECIPIENT);
+
+        awaitAtMostOneMinute.until(this::messageIsReceivedByTheSmtpServer);
+    }
+
+    @Test
+    public void directResolutionShouldFailoverOnSecondMxWhenFirstMxFailed() throws Exception {
+        InMemoryDNSService inMemoryDNSService = new InMemoryDNSService()
+            .registerRecord(JAMES_ANOTHER_DOMAIN, ImmutableList.of(), JAMES_ANOTHER_MX_DOMAINS, ImmutableList.of())
+            .registerMxRecord(JAMES_ANOTHER_MX_DOMAIN_1, LOCALHOST_IP) // TODO find another port
```
Comment (trantienduchn) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java]: `LOCALHOST_IP` is not good. I need suggestion :thinking:

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            .sendMessage(FROM, RECIPIENT);
+
+        awaitAtMostOneMinute.until(this::messageIsReceivedByTheSmtpServer);
+    }
+
+    @Test
+    public void directResolutionShouldFailoverOnSecondMxWhenFirstMxFailed() throws Exception {
+        InMemoryDNSService inMemoryDNSService = new InMemoryDNSService()
+            .registerRecord(JAMES_ANOTHER_DOMAIN, ImmutableList.of(), JAMES_ANOTHER_MX_DOMAINS, ImmutableList.of())
+            .registerMxRecord(JAMES_ANOTHER_MX_DOMAIN_1, LOCALHOST_IP) // TODO find another IP address
```
Comment (aduprat) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java]: You can introduce a second FakeSmtp rule.

Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java]: Or launch a container that do not listen on port 25 ?

Comment (mbaechler) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java]: you should choose an unroutable address like 10.x or 192.168.x

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        messageSender.connect(LOCALHOST_IP, SMTP_PORT)
+            .sendMessage(FROM, RECIPIENT);
+
+        awaitAtMostOneMinute.until(this::messageIsReceivedByTheSmtpServer);
+    }
+
+    @Test
+    public void directResolutionShouldBounceUponUnreachableMxRecords() throws Exception {
+        InMemoryDNSService inMemoryDNSService = new InMemoryDNSService()
+            .registerRecord(JAMES_ANOTHER_DOMAIN, ImmutableList.of(), ImmutableList.of("unknown"), ImmutableList.of());
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java]: could you name the empty lists please ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    private boolean messageIsReceivedByTheSmtpServer() {
+        return fakeSmtp.isReceived(response -> response
+            .body("", hasSize(1))
+            .body("[0].from", equalTo(FROM))
+            .body("[0].subject", equalTo("test")));
+    }
+
+    private ProcessorConfiguration.Builder directResolutionTransport() {
+        return ProcessorConfiguration.transport()
+            .addMailet(MailetConfiguration.BCC_STRIPPER)
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java]: I think this line is useless for these tests

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+            .addMailet(MailetConfiguration.remoteDeliveryBuilder()
+                .matcher(All.class));
+    }
+
+    private ProcessorConfiguration.Builder transport() {
+        return ProcessorConfiguration.transport()
+            .addMailet(MailetConfiguration.BCC_STRIPPER)
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java]: same here

**Comment by aduprat** [comment]
Build triggered. sha1 is merged.

**Comment by aduprat** [comment]
Build started sha1 is merged.

**Comment by aduprat** [comment]
Build finished. No test results found.

**Comment by aduprat** [comment]
Build triggered. sha1 is merged.

**Comment by aduprat** [comment]
Build started sha1 is merged.

**Comment by aduprat** [comment]
Build finished. No test results found.

**Comment by aduprat** [comment]
Build triggered. sha1 is merged.

**Comment by aduprat** [comment]
Build started sha1 is merged.

**Comment by aduprat** [comment]
Build finished. No test results found.

**Comment by aduprat** [comment]
Build triggered. sha1 is merged.

**Comment by aduprat** [comment]
Build started sha1 is merged.

**Comment by aduprat** [comment]
Build finished. No test results found.

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java (hunk 4 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java]: It's compulsory for Guice based mailet container (default check)

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java (hunk 5 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java]: It's compulsory for Guice based mailet container (default check)

**Comment by aduprat** [comment]
test this please

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    private static final String JAMES_ANOTHER_DOMAIN = "james.com";
+    private static final String JAMES_ANOTHER_MX_DOMAIN_1 = "mx1.james.com";
+    private static final String JAMES_ANOTHER_MX_DOMAIN_2 = "mx2.james.com";
+    private static final List<String> JAMES_ANOTHER_MX_DOMAINS = ImmutableList.of(JAMES_ANOTHER_MX_DOMAIN_1, JAMES_ANOTHER_MX_DOMAIN_2);
+
+    private static final String FROM = "from@" + DEFAULT_DOMAIN;
+    private static final String RECIPIENT = "touser@" + JAMES_ANOTHER_DOMAIN;
+
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
```
[the remark this answers is not in the document]
Comment (trantienduchn) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/DirectResolutionRemoteDeliveryIntegrationTest.java]: For now I don't know how to migrate Junit 4 Rules to Junit 5 TemporaryFolderExtension. And other rules (IMAPMessageReader, SMTPMessageSender, FakeSmtp - as I know testcontainer supports only rule, not extension). Migrate all rules to Junit 5 may be  overwhelmed to me 😄

**Comment by trantienduchn** [comment]
I just did some easy part of this issue. There are to go remaining tests:
```
- 5XX Errors don't trigger failover and a bounce is directly sent
- 4XX codes don't trigger failover, and will be retried. Moreover they can succeed on second try, generating no bounce.
- 4XX codes, if answered on each try, will lead to a bounce
Partial failure (some recipients are failing, some success):
   - Only failed (but valid) recipients should be retried (EG 450) and can suddeed on second try
   - Only failed (but valid) recipients should be retried (EG 450) and will generate a bounce if repeated
   - Invalid recipients (EG 550) directly generate a bounce, and will not be retried
   - Valid, and well delivered recipients wont be re-delivered
```

File: server/testing/src/main/java/org/apache/james/utils/FakeSmtp.java [hunk 7]
```diff
@@ -42,15 +42,32 @@
...
     private static final ResponseSpecification RESPONSE_SPECIFICATION = new ResponseSpecBuilder().build();
     private final SwarmGenericContainer container;
+    private final Integer smtpPort;
 
     public FakeSmtp() {
-        container = new SwarmGenericContainer(Images.FAKE_SMTP)
-            .withExposedPorts(SMTP_PORT)
-            .withAffinityToContainer()
-            .waitingFor(new HostPortWaitStrategy());
+        this(new SwarmGenericContainer(Images.FAKE_SMTP)
```
Comment (chibenwa) [inline server/testing/src/main/java/org/apache/james/utils/FakeSmtp.java]: Container declaration seems to be duplicating withSmtpPort

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**